## <center> Packages & Connection </center>

In [1]:
import os
import h3 as h3
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from scipy import stats
from pyhive import presto
from keplergl import KeplerGl
from datetime import datetime, timedelta

import warnings
warnings.filterwarnings('ignore')

In [2]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 500)

In [3]:
## Connection
connection = presto.connect(
        host='presto-gateway.serving.data.production.internal',
        port=80,
        protocol='http',
        catalog='hive',
        username='manoj.ravirajan@rapido.bike'
)

## Parameter

In [4]:
city = 'Bangalore'
service = 'Link'

In [5]:
start_date = '20230717'
end_date = '20230813'

In [7]:
## experiments.blr_use_case_hex12

blr_use_case = f"""

       WITH base_data AS (

        SELECT  
            yyyymmdd,
            city_name,
            service_obj_service_name,
            COALESCE(merged_usecase_accuracy, 'Unknown') usecase,
            pickup_location_hex_8,
            pickup_location_hex_12,
            order_id,
            order_status,
            spd_fraud_flag
        
        FROM
            orders.order_logs_snapshot ols 
        
        LEFT JOIN 
            experiments.blr_use_case_hex12
            ON pickup_location_hex_12 = hex_12
        
        WHERE 
            yyyymmdd >= '{start_date}'
            AND yyyymmdd <= '{end_date}'
            AND city_name = '{city}'
            AND service_obj_service_name = '{service}'
    ),
    
    hex_8_agg AS (

        SELECT 
            city_name,
            pickup_location_hex_8,
            usecase,
            COUNT(DISTINCT pickup_location_hex_12) total_pickup_hex_12,
            COUNT(DISTINCT order_id) gross_orders,
            COUNT(DISTINCT CASE WHEN order_status = 'dropped' AND (spd_fraud_flag = false OR spd_fraud_flag IS NULL) THEN order_id END) net_orders
        FROM
            base_data
        -- WHERE 
            -- pickup_location_hex_8 = '8860145a0dfffff'
        GROUP BY 1,2,3
    ),

    view AS (
    
    SELECT
        *,
        LEAD(total_pickup_hex_12,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY total_pickup_hex_12 DESC) next_hex_count_geo,
        LEAD(usecase,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY total_pickup_hex_12 DESC) next_use_case_geo,
        LEAD(gross_orders,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY total_pickup_hex_12 DESC) next_gross_orders_geo,
        LEAD(net_orders,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY total_pickup_hex_12 DESC) next_net_orders_geo,

        LEAD(total_pickup_hex_12,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY gross_orders DESC) next_hex_count_gross,
        LEAD(usecase,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY gross_orders DESC) next_use_case_gross,
        LEAD(gross_orders,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY gross_orders DESC) next_gross_orders_gross,
        LEAD(net_orders,1) OVER (PARTITION BY city_name,pickup_location_hex_8 ORDER BY gross_orders DESC) next_net_orders_gross,

        ROW_NUMBER() OVER(PARTITION BY city_name,pickup_location_hex_8 ORDER BY total_pickup_hex_12 DESC) geo_hexes_rn_desc,
        ROW_NUMBER() OVER(PARTITION BY city_name,pickup_location_hex_8 ORDER BY total_pickup_hex_12 ASC) geo_hexes_rn_asc,
        ROW_NUMBER() OVER(PARTITION BY city_name,pickup_location_hex_8 ORDER BY gross_orders DESC) gross_orders_rn_desc,
        ROW_NUMBER() OVER(PARTITION BY city_name,pickup_location_hex_8 ORDER BY gross_orders ASC) gross_orders_rn_asc
    
    FROM 
        hex_8_agg
    ),
    
    hex_agg AS (
    
    SELECT 
        flag,
        pickup_location_hex_8,
        SUM(total_pickup_hex_12) total_pickup_hex_12,
        SUM(gross_orders) total_gross_orders,
        SUM(net_orders) total_net_orders
    FROM 
    (
        SELECT 
            'geo_level_hex_tag' AS flag,
            pickup_location_hex_8,
            usecase,
            total_pickup_hex_12,
            gross_orders,
            net_orders,
            CASE 
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_asc = 1 THEN 'yes'
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_desc != geo_hexes_rn_asc THEN 'no'
            WHEN usecase != 'Unknown' THEN 'yes'
            END filter 
        FROM 
            view 
            
        UNION ALL 
        
        SELECT 
            'gross_level_hex_tag' AS flag,
            pickup_location_hex_8,
            usecase,
            total_pickup_hex_12,
            gross_orders,
            net_orders,
            CASE 
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_asc = 1 THEN 'yes'
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_desc != gross_orders_rn_asc THEN 'no'
            WHEN usecase != 'Unknown' THEN 'yes'
            END filter 
        FROM
            view 
    ) 
    WHERE filter = 'yes'
    GROUP BY 1,2
    )
    
    SELECT 
        a.flag,
        a.city_name,
        a.pickup_location_hex_8,
        a.usecase,
        a.usecase_pickup_hex_12,
        hex_agg.total_pickup_hex_12,
        COALESCE(TRY(a.usecase_pickup_hex_12*100.00/hex_agg.total_pickup_hex_12),0)  accuracy,
        a.usecase_gross_orders,
        hex_agg.total_gross_orders,
        a.usecase_net_orders,
        hex_agg.total_net_orders
    
    FROM 
    (
        SELECT 
            'geo_level_hex_tag' AS flag,
            city_name,
            pickup_location_hex_8,
            CASE 
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_asc = 1 THEN usecase
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_desc != geo_hexes_rn_asc THEN next_use_case_geo
            WHEN usecase != 'Unknown' THEN usecase
            END usecase,
            CASE 
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_asc = 1 THEN total_pickup_hex_12
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_desc != geo_hexes_rn_asc THEN next_hex_count_geo
            WHEN usecase != 'Unknown' THEN total_pickup_hex_12
            END usecase_pickup_hex_12,
            CASE 
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_asc = 1 THEN gross_orders
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_desc != geo_hexes_rn_asc THEN next_gross_orders_geo
            WHEN usecase != 'Unknown' THEN gross_orders
            END usecase_gross_orders,
            
            CASE 
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_asc = 1 THEN net_orders
            WHEN usecase = 'Unknown' AND geo_hexes_rn_desc = 1 AND geo_hexes_rn_desc != geo_hexes_rn_asc THEN next_net_orders_geo
            WHEN usecase != 'Unknown' THEN net_orders
            END usecase_net_orders

        FROM 
            view
        WHERE 
            geo_hexes_rn_desc = 1
            
        UNION ALL 
        
        SELECT 
            'gross_level_hex_tag' AS flag,
            city_name,
            pickup_location_hex_8,
            CASE 
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_asc = 1 THEN usecase
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_desc != gross_orders_rn_asc THEN next_use_case_gross
            WHEN usecase != 'Unknown' THEN usecase
            END usecase,
            CASE 
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_asc = 1 THEN total_pickup_hex_12
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_desc != gross_orders_rn_asc THEN next_hex_count_gross
            WHEN usecase != 'Unknown' THEN total_pickup_hex_12
            END usecase_pickup_hex_12,
            CASE 
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_asc = 1 THEN gross_orders
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_desc != gross_orders_rn_asc THEN next_gross_orders_gross
            WHEN usecase != 'Unknown' THEN gross_orders
            END usecase_gross_orders,
            
            CASE 
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_asc = 1 THEN net_orders
            WHEN usecase = 'Unknown' AND gross_orders_rn_desc = 1 AND gross_orders_rn_desc != gross_orders_rn_asc THEN next_net_orders_gross
            WHEN usecase != 'Unknown' THEN net_orders
            END usecase_net_orders


        FROM 
            view
        WHERE 
            gross_orders_rn_desc = 1
    ) AS a 
    
    LEFT JOIN 
        hex_agg
        ON a.flag = hex_agg.flag
        AND a.pickup_location_hex_8 = hex_agg.pickup_location_hex_8

"""

In [100]:
df_blr_use_case = pd.read_sql(blr_use_case, connection)
df_blr_use_case.head(3)

flag  city_name pickup_location_hex_8      usecase  \
0  gross_level_hex_tag  Bangalore       88601459b7fffff      Unknown   
1  gross_level_hex_tag  Bangalore       88601696b9fffff      Unknown   
2  gross_level_hex_tag  Bangalore       88618920b1fffff  residential   

   usecase_pickup_hex_12  total_pickup_hex_12  accuracy  usecase_gross_orders  \
0                     41                   41    100.00                    56   
1                      3                    3    100.00                     5   
2                    564                 1063     53.06                  6971   

   total_gross_orders  usecase_net_orders  total_net_orders  
0                  56                  12                12  
1                   5                   0                 0  
2               22588                2988             10274

In [ ]:
df_blr_use_case.to_csv('/Users/rapido/local-datasets/affluence/raw/hex12_residential_raw_data_{}_{}_{}_to_{}.csv' \
                                .format(city, service,start_date,end_date)
                               , index = False)

In [6]:
df_blr_use_case = pd.read_csv('/Users/rapido/local-datasets/affluence/raw/hex12_residential_raw_data_{}_{}_{}_to_{}.csv' \
                               .format(city, service,start_date,end_date))

In [7]:
df_use_case = df_blr_use_case.copy(deep=True)
print(df_use_case.shape)

(4800, 11)


In [8]:
df_hex_affluence_tag = pd.read_csv('/Users/rapido/local-datasets/affluence/main/hex_affluence_tag.csv')
df_hex_affluence_tag = df_hex_affluence_tag[['pickup_hex_8', 'affluence_tag']]

df_aff_use_case = pd.merge(df_use_case, 
                       df_hex_affluence_tag,
                       how = 'left',
                       left_on = ['pickup_location_hex_8'],
                       right_on = ['pickup_hex_8']
                       )


df_aff_use_case = df_aff_use_case[['affluence_tag', 'flag', 'city_name', 'pickup_location_hex_8', 'usecase',
                                   'usecase_pickup_hex_12', 'total_pickup_hex_12', 'accuracy',
                                   'usecase_gross_orders', 'total_gross_orders', 'usecase_net_orders',
                                   'total_net_orders']]
df_aff_use_case.head()

affluence_tag                 flag  city_name pickup_location_hex_8  \
0          Less  gross_level_hex_tag  Bangalore       88601459b7fffff   
1          High  gross_level_hex_tag  Bangalore       88601696b9fffff   
2          High  gross_level_hex_tag  Bangalore       88618920b1fffff   
3          Less  gross_level_hex_tag  Bangalore       8861892707fffff   
4          Less    geo_level_hex_tag  Bangalore       88601459b7fffff   

       usecase  usecase_pickup_hex_12  total_pickup_hex_12  accuracy  \
0      Unknown                     41                   41    100.00   
1      Unknown                      3                    3    100.00   
2  residential                    564                 1063     53.06   
3      Unknown                    226                  226    100.00   
4      Unknown                     41                   41    100.00   

   usecase_gross_orders  total_gross_orders  usecase_net_orders  \
0                    56                  56                  12   
1                     5                   5                   0   
2                  6971               22588                2988   
3                   593                 593                 266   
4                    56                  56                  12   

   total_net_orders  
0                12  
1                 0  
2             10274  
3               266  
4                12

In [9]:
df_aff_use_case['residential'] = np.where(df_aff_use_case['usecase'].isin(['residential', 'Unknown']), 
                                        df_aff_use_case['usecase'],
                                        'non-residential' )

In [10]:
df_aff_use_case.flag.unique()

array(['gross_level_hex_tag', 'geo_level_hex_tag'], dtype=object)

In [11]:
df_aff_use_case[df_aff_use_case['pickup_location_hex_8'] == '8860145865fffff']

affluence_tag                 flag  city_name pickup_location_hex_8  \
1574          High  gross_level_hex_tag  Bangalore       8860145865fffff   
1656          High    geo_level_hex_tag  Bangalore       8860145865fffff   

          usecase  usecase_pickup_hex_12  total_pickup_hex_12  accuracy  \
1574      leisure                     63                  252     25.00   
1656  residential                     70                  252     27.78   

      usecase_gross_orders  total_gross_orders  usecase_net_orders  \
1574                   206                 881                 133   
1656                   200                 881                 131   

      total_net_orders      residential  
1574               586  non-residential  
1656               586      residential

In [13]:
df_home_hex_8 = pd.read_clipboard()
df_home_hex_8

hex_8 residence_tag
0     8861892743fffff          Home
1     88618920d7fffff          Home
2     8860169157fffff          Home
3     886016974bfffff          Home
4     8861892509fffff          Home
...               ...           ...
1661  88618925c7fffff          Home
1662  8861892f59fffff          Home
1663  8860145827fffff          Home
1664  8861892417fffff          Home
1665  8860145321fffff          Home

[1666 rows x 2 columns]

In [14]:
df_aff_use_case = pd.merge(df_aff_use_case,
                           df_home_hex_8,
                           how = 'left',
                           left_on = ['pickup_location_hex_8'],
                           right_on = ['hex_8']
                          )
df_aff_use_case                           

affluence_tag                 flag  city_name pickup_location_hex_8  \
0             Less  gross_level_hex_tag  Bangalore       88601459b7fffff   
1             High  gross_level_hex_tag  Bangalore       88601696b9fffff   
2             High  gross_level_hex_tag  Bangalore       88618920b1fffff   
3             Less  gross_level_hex_tag  Bangalore       8861892707fffff   
4             Less    geo_level_hex_tag  Bangalore       88601459b7fffff   
...            ...                  ...        ...                   ...   
4795          Less  gross_level_hex_tag  Bangalore       88618922cbfffff   
4796          High  gross_level_hex_tag  Bangalore       8861892439fffff   
4797          High    geo_level_hex_tag  Bangalore       8860144ae5fffff   
4798          High    geo_level_hex_tag  Bangalore       88618920d3fffff   
4799          Less    geo_level_hex_tag  Bangalore       8861892aa1fffff   

          usecase  usecase_pickup_hex_12  total_pickup_hex_12  accuracy  \
0         Unknown                     41                   41    100.00   
1         Unknown                      3                    3    100.00   
2     residential                    564                 1063     53.06   
3         Unknown                    226                  226    100.00   
4         Unknown                     41                   41    100.00   
...           ...                    ...                  ...       ...   
4795      Unknown                     44                   44    100.00   
4796  residential                    499                  863     57.82   
4797      Unknown                      2                    2    100.00   
4798  residential                     11                   11    100.00   
4799      Unknown                    213                  213    100.00   

      usecase_gross_orders  total_gross_orders  usecase_net_orders  \
0                       56                  56                  12   
1                        5                   5                   0   
2                     6971               22588                2988   
3                      593                 593                 266   
4                       56                  56                  12   
...                    ...                 ...                 ...   
4795                    76                  76                  22   
4796                  3922                7511                1509   
4797                     2                   2                   0   
4798                    27                  27                  11   
4799                   784                 784                 285   

      total_net_orders  residential            hex_8 residence_tag  
0                   12      Unknown  88601459b7fffff          Home  
1                    0      Unknown  88601696b9fffff          Home  
2                10274  residential  88618920b1fffff          Home  
3                  266      Unknown  8861892707fffff          Home  
4                   12      Unknown  88601459b7fffff          Home  
...                ...          ...              ...           ...  
4795                22      Unknown              NaN           NaN  
4796              3066  residential  8861892439fffff          Home  
4797                 0      Unknown              NaN           NaN  
4798                11  residential  88618920d3fffff          Home  
4799               285      Unknown  8861892aa1fffff          Home  

[4800 rows x 15 columns]

In [20]:
df_aff_use_case.to_csv('/Users/rapido/local-datasets/affluence/final/geo_use_case_all_hex_8.csv', index = False)

In [15]:
df_aff_use_case.to_csv('/Users/rapido/local-datasets/affluence/final/geo_use_case_all_hex_81.csv', index = False)

## Analysis

### Approach 1 
    -- We can consider primary(mode/most repeated use case) hex_12 use_case for the respective hex 8
        - Disadvantage -> Geo-based tagging
        
### Approach 2
    -- We can consider hex_12 use_case that's having more orders contribution for the respective hex 8
        - Disadvantage -> Orders based tagging

In [16]:
df_aff_agg = df_aff_use_case \
            .groupby(['affluence_tag', 'flag']) \
            .agg( aff_hex_count = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                  aff_gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                  aff_net_orders = pd.NamedAgg('total_net_orders', 'sum')
                ).reset_index()
df_aff_agg

affluence_tag                 flag  aff_hex_count  aff_gross_orders  \
0          High    geo_level_hex_tag           1045           2981551   
1          High  gross_level_hex_tag           1045           2981551   
2          Less    geo_level_hex_tag           1205            341655   
3          Less  gross_level_hex_tag           1205            341655   

   aff_net_orders  
0         1514059  
1         1514059  
2          159593  
3          159593

In [17]:
## Approach 1

df_aff_use_case_type1 = df_aff_use_case[df_aff_use_case['flag'] == 'geo_level_hex_tag']
df_aff_use_case_type1.head(2)

affluence_tag               flag  city_name pickup_location_hex_8  usecase  \
4          Less  geo_level_hex_tag  Bangalore       88601459b7fffff  Unknown   
5          High  geo_level_hex_tag  Bangalore       886014435bfffff  Unknown   

   usecase_pickup_hex_12  total_pickup_hex_12  accuracy  usecase_gross_orders  \
4                     41                   41     100.0                    56   
5                     12                   12     100.0                    18   

   total_gross_orders  usecase_net_orders  total_net_orders residential  \
4                  56                  12                12     Unknown   
5                  18                   0                 0     Unknown   

             hex_8 residence_tag  
4  88601459b7fffff          Home  
5              NaN           NaN

In [18]:
df_type1_analysis = df_aff_use_case_type1 \
                            .groupby(['affluence_tag', 'usecase']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type1_analysis['city_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis.no_hexes.sum()).round(2)
df_type1_analysis['city_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis.gross_orders.sum()).round(2)
df_type1_analysis['agg_hexes'] = np.where(df_type1_analysis['affluence_tag'] == 'High' ,
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type1_analysis['agg_gross'] = np.where(df_type1_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type1_analysis['affluence_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis['agg_hexes']).round(2)
df_type1_analysis['affluence_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis['agg_gross']).round(2)


df_type1_analysis[['affluence_tag', 'usecase', #'residence_tag',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'affluence_gross_distr'], ascending = False)

affluence_tag              usecase  city_hex_distr  affluence_hex_distr  \
20          Less          residential            7.11                13.28   
11          Less              Unknown           41.64                77.76   
21          Less      transit_station            0.58                 1.08   
12          Less          educational            1.42                 2.66   
18          Less               office            0.76                 1.41   
17          Less              leisure            0.89                 1.66   
16          Less      household_needs            0.40                 0.75   
15          Less  health_and_personal            0.31                 0.58   
13          Less                 food            0.27                 0.50   
19          Less     place_of_worship            0.13                 0.25   
14          Less         govt_amenity            0.04                 0.08   
9           High          residential           21.78                46.89   
7           High               office            2.22                 4.78   
6           High              leisure            3.38                 7.27   
0           High              Unknown           14.80                31.87   
5           High      household_needs            1.24                 2.68   
1           High          educational            1.51                 3.25   
2           High                 food            0.40                 0.86   
10          High      transit_station            0.44                 0.96   
4           High  health_and_personal            0.58                 1.24   
3           High         govt_amenity            0.04                 0.10   
8           High     place_of_worship            0.04                 0.10   

    city_gross_distr  affluence_gross_distr  
20              3.55                  34.50  
11              2.71                  26.39  
21              1.42                  13.86  
12              0.64                   6.21  
18              0.60                   5.83  
17              0.46                   4.44  
16              0.43                   4.16  
15              0.42                   4.08  
13              0.03                   0.32  
19              0.02                   0.16  
14              0.01                   0.06  
9              56.96                  63.48  
7              11.01                  12.27  
6              10.20                  11.37  
0               4.02                   4.48  
5               2.71                   3.02  
1               2.35                   2.62  
2               0.89                   0.99  
10              0.75                   0.84  
4               0.75                   0.83  
3               0.06                   0.07  
8               0.02                   0.02

kepler to see the 55 % of unknown hex 8's 

### Insights 

####  Less affluence
           
           residential - 13% hex's contributing  34.5% gross 
           unknown - 77.76% hex's contributing  26% gross
           transit_station - 1.08% hex's contributing  13% gross
            
####    High affluence
            
            residential - 46.89% hex's contributing  63.48% gross 
            office - 4.78% hex's contributing  12% gross
            leisure - 7.27% hex's contributing  11% gross  

In [19]:
df_type1_analysis = df_aff_use_case_type1 \
                            .groupby(['affluence_tag', 'residential']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type1_analysis['city_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis.no_hexes.sum()).round(2)
df_type1_analysis['city_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis.gross_orders.sum()).round(2)
df_type1_analysis['agg_hexes'] = np.where(df_type1_analysis['affluence_tag'] == 'High' ,
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type1_analysis['agg_gross'] = np.where(df_type1_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type1_analysis['affluence_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis['agg_hexes']).round(2)
df_type1_analysis['affluence_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis['agg_gross']).round(2)


df_type1_analysis[['affluence_tag', 'residential', 'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'residential'], ascending = False)

affluence_tag      residential  no_hexes  city_hex_distr  \
5          Less      residential       160            7.11   
4          Less  non-residential       108            4.80   
3          Less          Unknown       937           41.64   
2          High      residential       490           21.78   
1          High  non-residential       222            9.87   
0          High          Unknown       333           14.80   

   affluence_hex_distr  city_gross_distr  affluence_gross_distr  
5                13.28              3.55                  34.50  
4                 8.96              4.02                  39.11  
3                77.76              2.71                  26.39  
2                46.89             56.96                  63.48  
1                21.24             28.74                  32.03  
0                31.87              4.02                   4.48

In [20]:
df_type1_analysis = df_aff_use_case_type1 \
                            .groupby(['affluence_tag']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type1_analysis['city_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis.no_hexes.sum()).round(2)
df_type1_analysis['city_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis.gross_orders.sum()).round(2)
df_type1_analysis['agg_hexes'] = np.where(df_type1_analysis['affluence_tag'] == 'High' ,
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type1_analysis['agg_gross'] = np.where(df_type1_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type1_analysis['affluence_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis['agg_hexes']).round(2)
df_type1_analysis['affluence_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis['agg_gross']).round(2)


df_type1_analysis[['affluence_tag', 'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag'], ascending = False)

affluence_tag  no_hexes  city_hex_distr  affluence_hex_distr  \
1          Less      1205           53.56                100.0   
0          High      1045           46.44                100.0   

   city_gross_distr  affluence_gross_distr  
1             10.28                  100.0  
0             89.72                  100.0

## Logic

In [21]:
df_aff_use_case_type1['city_geo_flag'] = np.where(df_aff_use_case_type1['usecase'] == 'Unknown', 'Outskirts', 'Intra City')
df_aff_use_case_type1['density_flag'] = np.where(df_aff_use_case_type1['total_gross_orders'] >= df_aff_use_case_type1.total_gross_orders.quantile(0.76) ,  'High', 'Less')

df_aff_use_case_type1

affluence_tag               flag  city_name pickup_location_hex_8  \
4             Less  geo_level_hex_tag  Bangalore       88601459b7fffff   
5             High  geo_level_hex_tag  Bangalore       886014435bfffff   
6             Less  geo_level_hex_tag  Bangalore       8860144a59fffff   
7             High  geo_level_hex_tag  Bangalore       88601451c1fffff   
8             Less  geo_level_hex_tag  Bangalore       88601451d7fffff   
...            ...                ...        ...                   ...   
4775          High  geo_level_hex_tag  Bangalore       8860145b05fffff   
4776          High  geo_level_hex_tag  Bangalore       886189273dfffff   
4797          High  geo_level_hex_tag  Bangalore       8860144ae5fffff   
4798          High  geo_level_hex_tag  Bangalore       88618920d3fffff   
4799          Less  geo_level_hex_tag  Bangalore       8861892aa1fffff   

          usecase  usecase_pickup_hex_12  total_pickup_hex_12  accuracy  \
4         Unknown                     41                   41    100.00   
5         Unknown                     12                   12    100.00   
6         Unknown                      1                    1    100.00   
7         Unknown                      3                    3    100.00   
8         Unknown                     77                   77    100.00   
...           ...                    ...                  ...       ...   
4775      leisure                    372                  677     54.95   
4776  residential                    168                  252     66.67   
4797      Unknown                      2                    2    100.00   
4798  residential                     11                   11    100.00   
4799      Unknown                    213                  213    100.00   

      usecase_gross_orders  total_gross_orders  usecase_net_orders  \
4                       56                  56                  12   
5                       18                  18                   0   
6                        2                   2                   0   
7                        5                   5                   0   
8                      235                 235                  55   
...                    ...                 ...                 ...   
4775                  1427                3529                 881   
4776                   764                1328                 324   
4797                     2                   2                   0   
4798                    27                  27                  11   
4799                   784                 784                 285   

      total_net_orders      residential            hex_8 residence_tag  \
4                   12          Unknown  88601459b7fffff          Home   
5                    0          Unknown              NaN           NaN   
6                    0          Unknown              NaN           NaN   
7                    0          Unknown              NaN           NaN   
8                   55          Unknown  88601451d7fffff          Home   
...                ...              ...              ...           ...   
4775              2129  non-residential  8860145b05fffff          Home   
4776               606      residential  886189273dfffff          Home   
4797                 0          Unknown              NaN           NaN   
4798                11      residential  88618920d3fffff          Home   
4799               285          Unknown  8861892aa1fffff          Home   

     city_geo_flag density_flag  
4        Outskirts         Less  
5        Outskirts         Less  
6        Outskirts         Less  
7        Outskirts         Less  
8        Outskirts         Less  
...            ...          ...  
4775    Intra City         High  
4776    Intra City         High  
4797     Outskirts         Less  
4798    Intra City         Less  
4799     Outskirts         Less  

[2400 rows x 17 columns]

In [22]:
df_type1_analysis = df_aff_use_case_type1 \
                            .groupby(['affluence_tag', 'density_flag']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type1_analysis['city_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis.no_hexes.sum()).round(2)
df_type1_analysis['city_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis.gross_orders.sum()).round(2)
df_type1_analysis['agg_hexes'] = np.where(df_type1_analysis['affluence_tag'] == 'High' ,
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type1_analysis['agg_gross'] = np.where(df_type1_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type1_analysis['affluence_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis['agg_hexes']).round(2)
df_type1_analysis['affluence_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis['agg_gross']).round(2)


df_type1_analysis[['affluence_tag', 'density_flag', 'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'density_flag'], ascending = False)

affluence_tag density_flag  no_hexes  city_hex_distr  affluence_hex_distr  \
3          Less         Less      1138           50.58                94.44   
2          Less         High        67            2.98                 5.56   
1          High         Less       536           23.82                51.29   
0          High         High       509           22.62                48.71   

   city_gross_distr  affluence_gross_distr  
3              3.99                  38.83  
2              6.29                  61.17  
1              3.58                   3.99  
0             86.14                  96.01

In [23]:
df_type1_analysis = df_aff_use_case_type1 \
                            .groupby(['affluence_tag', 'density_flag', 'city_geo_flag']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type1_analysis['city_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis.no_hexes.sum()).round(2)
df_type1_analysis['city_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis.gross_orders.sum()).round(2)
df_type1_analysis['agg_hexes'] = np.where(df_type1_analysis['affluence_tag'] == 'High' ,
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type1_analysis['agg_gross'] = np.where(df_type1_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type1_analysis['affluence_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis['agg_hexes']).round(2)
df_type1_analysis['affluence_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis['agg_gross']).round(2)


df_type1_analysis[['affluence_tag', 'density_flag', 'city_geo_flag', 'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'city_geo_flag'], ascending = False)

affluence_tag density_flag city_geo_flag  no_hexes  city_hex_distr  \
5          Less         High     Outskirts        10            0.44   
7          Less         Less     Outskirts       927           41.20   
4          Less         High    Intra City        57            2.53   
6          Less         Less    Intra City       211            9.38   
1          High         High     Outskirts        31            1.38   
3          High         Less     Outskirts       302           13.42   
0          High         High    Intra City       478           21.24   
2          High         Less    Intra City       234           10.40   

   affluence_hex_distr  city_gross_distr  affluence_gross_distr  
5                 0.83              0.49                   4.78  
7                76.93              2.22                  21.61  
4                 4.73              5.80                  56.39  
6                17.51              1.77                  17.22  
1                 2.97              3.23                   3.60  
3                28.90              0.79                   0.88  
0                45.74             82.91                  92.41  
2                22.39              2.79                   3.11

In [117]:
ppkm_test_df = pd.read_clipboard()
ppkm_test_df

affluence_tag     pickup_hex_8  ppkm_flag  fe_count
0             High  8860144107fffff  Less ppkm        22
1             High  8860144135fffff  Less ppkm        39
2             High  8860144145fffff  Less ppkm        35
3             High  886014414bfffff  Less ppkm        43
4             High  8860144227fffff  Less ppkm        44
...            ...              ...        ...       ...
2036          Less  88618935e3fffff  Less ppkm       458
2037          Less  88618935e7fffff  Less ppkm       258
2038          Less  88618935e9fffff  Less ppkm        22
2039          Less  8861893713fffff  Less ppkm       404
2040          Less  8861893893fffff  Less ppkm        68

[2041 rows x 4 columns]

In [118]:
df_final = pd.merge(df_aff_use_case_type1,
                    ppkm_test_df[['pickup_hex_8', 'ppkm_flag']],
                    how = 'left',
                    left_on = ['pickup_location_hex_8'],
                    right_on = ['pickup_hex_8'])
df_final['ppkm_flag'] = np.where(df_final['ppkm_flag'] == 'High ppkm', 'High ppkm', 
                                np.where(df_final['ppkm_flag'] == 'Less ppkm', 'Less ppkm', 'Less ppkm'
                                ))
df_final

affluence_tag               flag  city_name pickup_location_hex_8  \
0             Less  geo_level_hex_tag  Bangalore       88601459b7fffff   
1             High  geo_level_hex_tag  Bangalore       886014435bfffff   
2             Less  geo_level_hex_tag  Bangalore       8860144a59fffff   
3             High  geo_level_hex_tag  Bangalore       88601451c1fffff   
4             Less  geo_level_hex_tag  Bangalore       88601451d7fffff   
...            ...                ...        ...                   ...   
2395          High  geo_level_hex_tag  Bangalore       8860145b05fffff   
2396          High  geo_level_hex_tag  Bangalore       886189273dfffff   
2397          High  geo_level_hex_tag  Bangalore       8860144ae5fffff   
2398          High  geo_level_hex_tag  Bangalore       88618920d3fffff   
2399          Less  geo_level_hex_tag  Bangalore       8861892aa1fffff   

          usecase  usecase_pickup_hex_12  total_pickup_hex_12  accuracy  \
0         Unknown                     41                   41    100.00   
1         Unknown                     12                   12    100.00   
2         Unknown                      1                    1    100.00   
3         Unknown                      3                    3    100.00   
4         Unknown                     77                   77    100.00   
...           ...                    ...                  ...       ...   
2395      leisure                    372                  677     54.95   
2396  residential                    168                  252     66.67   
2397      Unknown                      2                    2    100.00   
2398  residential                     11                   11    100.00   
2399      Unknown                    213                  213    100.00   

      usecase_gross_orders  total_gross_orders  usecase_net_orders  \
0                       56                  56                  12   
1                       18                  18                   0   
2                        2                   2                   0   
3                        5                   5                   0   
4                      235                 235                  55   
...                    ...                 ...                 ...   
2395                  1427                3529                 881   
2396                   764                1328                 324   
2397                     2                   2                   0   
2398                    27                  27                  11   
2399                   784                 784                 285   

      total_net_orders      residential            hex_8 residence_tag  \
0                   12          Unknown  88601459b7fffff          Home   
1                    0          Unknown              NaN           NaN   
2                    0          Unknown              NaN           NaN   
3                    0          Unknown              NaN           NaN   
4                   55          Unknown  88601451d7fffff          Home   
...                ...              ...              ...           ...   
2395              2129  non-residential  8860145b05fffff          Home   
2396               606      residential  886189273dfffff          Home   
2397                 0          Unknown              NaN           NaN   
2398                11      residential  88618920d3fffff          Home   
2399               285          Unknown  8861892aa1fffff          Home   

     city_geo_flag density_flag     pickup_hex_8  ppkm_flag  
0        Outskirts         Less  88601459b7fffff  Less ppkm  
1        Outskirts         Less  886014435bfffff  Less ppkm  
2        Outskirts         Less              NaN  Less ppkm  
3        Outskirts         Less  88601451c1fffff  Less ppkm  
4        Outskirts         Less  88601451d7fffff  Less ppkm  
...            ...          ...              ...        ...  
2395    Intra City         High  8860145b05fffff  L

In [119]:
df_type1_analysis = df_final \
                            .groupby(['affluence_tag', 'density_flag', 'city_geo_flag', 'ppkm_flag']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type1_analysis['city_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis.no_hexes.sum()).round(2)
df_type1_analysis['city_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis.gross_orders.sum()).round(2)
df_type1_analysis['agg_hexes'] = np.where(df_type1_analysis['affluence_tag'] == 'High' ,
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type1_analysis['agg_gross'] = np.where(df_type1_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'geo_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type1_analysis['affluence_hex_distr'] = (df_type1_analysis['no_hexes']*100.0/df_type1_analysis['agg_hexes']).round(2)
df_type1_analysis['affluence_gross_distr'] = (df_type1_analysis['gross_orders']*100.0/df_type1_analysis['agg_gross']).round(2)


df_type1_analysis[['affluence_tag', 'density_flag', 'city_geo_flag', 'ppkm_flag', 'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'density_flag', 'city_geo_flag', 'ppkm_flag'], ascending = False)

affluence_tag density_flag city_geo_flag  ppkm_flag  no_hexes  \
14          Less         Less     Outskirts  Less ppkm       925   
13          Less         Less     Outskirts  High ppkm         2   
12          Less         Less    Intra City  Less ppkm       205   
11          Less         Less    Intra City  High ppkm         6   
10          Less         High     Outskirts  Less ppkm        10   
9           Less         High    Intra City  Less ppkm        51   
8           Less         High    Intra City  High ppkm         6   
7           High         Less     Outskirts  Less ppkm       296   
6           High         Less     Outskirts  High ppkm         6   
5           High         Less    Intra City  Less ppkm       230   
4           High         Less    Intra City  High ppkm         4   
3           High         High     Outskirts  Less ppkm        27   
2           High         High     Outskirts  High ppkm         4   
1           High         High    Intra City  Less ppkm       424   
0           High         High    Intra City  High ppkm        54   

    city_hex_distr  affluence_hex_distr  city_gross_distr  \
14           41.11                76.76              2.22   
13            0.09                 0.17              0.00   
12            9.11                17.01              1.71   
11            0.27                 0.50              0.06   
10            0.44                 0.83              0.49   
9             2.27                 4.23              5.16   
8             0.27                 0.50              0.64   
7            13.16                28.33              0.78   
6             0.27                 0.57              0.01   
5            10.22                22.01              2.72   
4             0.18                 0.38              0.06   
3             1.20                 2.58              2.77   
2             0.18                 0.38              0.46   
1            18.84                40.57             71.56   
0             2.40                 5.17             11.35   

    affluence_gross_distr  
14                  21.61  
13                   0.00  
12                  16.59  
11                   0.63  
10                   4.78  
9                   50.14  
8                    6.24  
7                    0.87  
6                    0.01  
5                    3.03  
4                    0.07  
3                    3.09  
2                    0.51  
1                   79.76  
0                   12.65

In [1]:
df_final_test = df_final[(df_final['affluence_tag'] == 'Less' )
                     &
                  (df_final['density_flag'] == 'High')
                    &
                  (df_final['city_geo_flag'] == 'Intra City')
                 ]
# df_final_test
df_final_test[['affluence_tag', 'density_flag', 'city_geo_flag', 'ppkm_flag', 'residential', 'pickup_location_hex_8']]

NameError: name 'df_final' is not defined

In [70]:
## canonical.clevertap_customer_fare_estimate

city_hex_8 = f"""
       SELECT 
            pic.cluster AS pickup_location,
            pic.hex_id AS hex_id
        FROM
            datasets.city_cluster_hex AS pic
        WHERE
            city = 'Bangalore'
            AND resolution = 8
       
"""

df_city_hex_8 = pd.read_sql(city_hex_8, connection)
df_city_hex_8

pickup_location           hex_id
0             Bidadi  8860144229fffff
1           Sarjapur  8861893515fffff
2        Nelamangala  8860144a2bfffff
3       South Bidadi  8860145089fffff
4        Frazer Town  8861892e97fffff
...              ...              ...
3225        Attibele  886014cb29fffff
3226        Attibele  88618934d9fffff
3227        Attibele  88618936a3fffff
3228        Attibele  886014cb65fffff
3229        Attibele  88618936adfffff

[3230 rows x 2 columns]

In [75]:
df_final_test_vw = pd.merge(df_final_test,
                            df_city_hex_8,
                            how = 'left',
                            left_on = ['pickup_location_hex_8'],
                            right_on = ['hex_id']
                           )
df_final_test_vw[['affluence_tag', 'density_flag', 'city_geo_flag', 'ppkm_flag', 'residential', 'pickup_location_hex_8', 'pickup_location']].sort_values(['affluence_tag', 'density_flag', 'city_geo_flag', 'ppkm_flag'], ascending = False)

affluence_tag density_flag city_geo_flag  ppkm_flag      residential  \
1           Less         High    Intra City  Less ppkm  non-residential   
8           Less         High    Intra City  Less ppkm      residential   
20          Less         High    Intra City  Less ppkm  non-residential   
27          Less         High    Intra City  Less ppkm  non-residential   
34          Less         High    Intra City  Less ppkm      residential   
0           Less         High    Intra City  High ppkm      residential   
2           Less         High    Intra City  High ppkm      residential   
3           Less         High    Intra City  High ppkm      residential   
4           Less         High    Intra City  High ppkm      residential   
5           Less         High    Intra City  High ppkm      residential   
6           Less         High    Intra City  High ppkm  non-residential   
7           Less         High    Intra City  High ppkm  non-residential   
9           Less         High    Intra City  High ppkm      residential   
10          Less         High    Intra City  High ppkm      residential   
11          Less         High    Intra City  High ppkm  non-residential   
12          Less         High    Intra City  High ppkm      residential   
13          Less         High    Intra City  High ppkm  non-residential   
14          Less         High    Intra City  High ppkm      residential   
15          Less         High    Intra City  High ppkm      residential   
16          Less         High    Intra City  High ppkm      residential   
17          Less         High    Intra City  High ppkm  non-residential   
18          Less         High    Intra City  High ppkm  non-residential   
19          Less         High    Intra City  High ppkm  non-residential   
21          Less         High    Intra City  High ppkm      residential   
22          Less         High    Intra City  High ppkm      residential   
23          Less         High    Intra City  High ppkm  non-residential   
24          Less         High    Intra City  High ppkm  non-residential   
25          Less         High    Intra City  High ppkm  non-residential   
26          Less         High    Intra City  High ppkm      residential   
28          Less         High    Intra City  High ppkm  non-residential   
29          Less         High    Intra City  High ppkm      residential   
30          Less         High    Intra City  High ppkm  non-residential   
31          Less         High    Intra City  High ppkm  non-residential   
32          Less         High    Intra City  High ppkm  non-residential   
33          Less         High    Intra City  High ppkm      residential   
35          Less         High    Intra City  High ppkm      residential   
36          Less         High    Intra City  High ppkm      residential   
37          Less         High    Intra City  High ppkm      residential   
38          Less         High    Intra City  High ppkm  non-residential   
39          Less         High    Intra City  High ppkm      residential   
40          Less         High    Intra City  High ppkm      residential   
41          Less         High    Intra City  High ppkm  non-residential   
42          Less         High    Intra City  High ppkm      residential   
43          Less         High    Intra City  High ppkm      residential   
44          Less         High    Intra City  High ppkm  non-residential   
45          Less         High    Intra City  High ppkm  non-residential   
46          Less         High    Intra City  High ppkm  non-residential   
47          Less         High    Intra City  High ppkm  non-residential   
48          Less         High    Intra City  High ppkm      residential   
49          Less         High    Intra City  High ppkm  non-residential   
50          Less         High    Intra City  High ppkm      residential   
51          Less         High    Intra City  High ppkm      residential   
52          Less         Hig

In [80]:
df_aff_use_case_type1

affluence_tag               flag  city_name pickup_location_hex_8  \
4             Less  geo_level_hex_tag  Bangalore       88601459b7fffff   
5             High  geo_level_hex_tag  Bangalore       886014435bfffff   
6             Less  geo_level_hex_tag  Bangalore       8860144a59fffff   
7             High  geo_level_hex_tag  Bangalore       88601451c1fffff   
8             Less  geo_level_hex_tag  Bangalore       88601451d7fffff   
...            ...                ...        ...                   ...   
4775          High  geo_level_hex_tag  Bangalore       8860145b05fffff   
4776          High  geo_level_hex_tag  Bangalore       886189273dfffff   
4797          High  geo_level_hex_tag  Bangalore       8860144ae5fffff   
4798          High  geo_level_hex_tag  Bangalore       88618920d3fffff   
4799          Less  geo_level_hex_tag  Bangalore       8861892aa1fffff   

          usecase  usecase_pickup_hex_12  total_pickup_hex_12  accuracy  \
4         Unknown                     41                   41    100.00   
5         Unknown                     12                   12    100.00   
6         Unknown                      1                    1    100.00   
7         Unknown                      3                    3    100.00   
8         Unknown                     77                   77    100.00   
...           ...                    ...                  ...       ...   
4775      leisure                    372                  677     54.95   
4776  residential                    168                  252     66.67   
4797      Unknown                      2                    2    100.00   
4798  residential                     11                   11    100.00   
4799      Unknown                    213                  213    100.00   

      usecase_gross_orders  total_gross_orders  usecase_net_orders  \
4                       56                  56                  12   
5                       18                  18                   0   
6                        2                   2                   0   
7                        5                   5                   0   
8                      235                 235                  55   
...                    ...                 ...                 ...   
4775                  1427                3529                 881   
4776                   764                1328                 324   
4797                     2                   2                   0   
4798                    27                  27                  11   
4799                   784                 784                 285   

      total_net_orders      residential            hex_8 residence_tag  \
4                   12          Unknown  88601459b7fffff          Home   
5                    0          Unknown              NaN           NaN   
6                    0          Unknown              NaN           NaN   
7                    0          Unknown              NaN           NaN   
8                   55          Unknown  88601451d7fffff          Home   
...                ...              ...              ...           ...   
4775              2129  non-residential  8860145b05fffff          Home   
4776               606      residential  886189273dfffff          Home   
4797                 0          Unknown              NaN           NaN   
4798                11      residential  88618920d3fffff          Home   
4799               285          Unknown  8861892aa1fffff          Home   

     city_geo_flag density_flag  
4        Outskirts         Less  
5        Outskirts         Less  
6        Outskirts         Less  
7        Outskirts         Less  
8        Outskirts         Less  
...            ...          ...  
4775    Intra City         High  
4776    Intra City         High  
4797     Outskirts         Less  
4798    Intra City         Less  
4799     Outskirts         Less  

[2400 rows x 17 columns]

In [97]:
bangalore_affluence_pickup_cluster_hex8 = pd.merge(df_aff_use_case_type1,
                                                   df_city_hex_8,
                                                   how='left',
                                                   left_on = ['pickup_location_hex_8'],
                                                   right_on = ['hex_id']
                                                  )
bangalore_affluence_pickup_cluster_hex8 = bangalore_affluence_pickup_cluster_hex8[['pickup_location', 'pickup_location_hex_8', 'affluence_tag']].sort_values(['affluence_tag'])
bangalore_affluence_pickup_cluster_hex8 = bangalore_affluence_pickup_cluster_hex8[bangalore_affluence_pickup_cluster_hex8['affluence_tag'].isin(['High', 'Less'])]
bangalore_affluence_pickup_cluster_hex8

pickup_location pickup_location_hex_8 affluence_tag
1976                       Nelamangala       8860144b41fffff          High
949            Electronic City Phase 2       8861892627fffff          High
948                   Kurudusonnehalli       88618921b5fffff          High
947   Munnekollal Sankara Eye Hospital       88618920bdfffff          High
946        Metro - Sandal Soap Factory       8860145b37fffff          High
...                                ...                   ...           ...
1038                       Chikkagobbi       8861892d47fffff          Less
1037                   Soladevanahalli       8860169699fffff          Less
1036                           Bagalur       88601692d1fffff          Less
1029                          Sarjapur       88618922d5fffff          Less
2399                       Hoskote_BLR       8861892aa1fffff          Less

[2250 rows x 3 columns]

In [98]:
bangalore_affluence_pickup_cluster_hex8.to_csv('/Users/rapido/local-datasets/affluence/main/bangalore_affluence_pickup_cluster_hex8.csv', \
                                              index=False
                                              )

## Approach 2

In [25]:
## Approach 2

df_aff_use_case_type2 = df_aff_use_case[df_aff_use_case['flag'] == 'gross_level_hex_tag']

In [26]:
df_type2_analysis = df_aff_use_case_type2 \
                            .groupby(['affluence_tag', 'usecase']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type2_analysis['city_hex_distr'] = (df_type2_analysis['no_hexes']*100.0/df_type2_analysis.no_hexes.sum()).round(2)
df_type2_analysis['city_gross_distr'] = (df_type2_analysis['gross_orders']*100.0/df_type2_analysis.gross_orders.sum()).round(2)
df_type2_analysis['agg_hexes'] = np.where(df_type2_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type2_analysis['agg_gross'] = np.where(df_type2_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type2_analysis['affluence_hex_distr'] = (df_type2_analysis['no_hexes']*100.0/df_type2_analysis['agg_hexes']).round(2)
df_type2_analysis['affluence_gross_distr'] = (df_type2_analysis['gross_orders']*100.0/df_type2_analysis['agg_gross']).round(2)


df_type2_analysis[['affluence_tag', 'usecase', #'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'affluence_gross_distr'], ascending = False)

affluence_tag              usecase  city_hex_distr  affluence_hex_distr  \
11          Less              Unknown           41.64                77.76   
21          Less      transit_station            1.11                 2.07   
20          Less          residential            6.58                12.28   
18          Less               office            0.98                 1.83   
16          Less      household_needs            0.36                 0.66   
12          Less          educational            1.42                 2.66   
17          Less              leisure            0.62                 1.16   
13          Less                 food            0.31                 0.58   
14          Less         govt_amenity            0.04                 0.08   
15          Less  health_and_personal            0.31                 0.58   
19          Less     place_of_worship            0.18                 0.33   
9           High          residential           18.93                40.77   
7           High               office            2.98                 6.41   
6           High              leisure            3.02                 6.51   
10          High      transit_station            2.09                 4.50   
2           High                 food            0.93                 2.01   
0           High              Unknown           14.80                31.87   
5           High      household_needs            1.42                 3.06   
1           High          educational            1.42                 3.06   
4           High  health_and_personal            0.76                 1.63   
3           High         govt_amenity            0.04                 0.10   
8           High     place_of_worship            0.04                 0.10   

    city_gross_distr  affluence_gross_distr  
11              2.71                  26.39  
21              2.59                  25.23  
20              2.55                  24.78  
18              0.86                   8.34  
16              0.48                   4.62  
12              0.46                   4.47  
17              0.32                   3.12  
13              0.15                   1.47  
14              0.09                   0.89  
15              0.05                   0.48  
19              0.02                   0.20  
9              44.37                  49.45  
7              14.64                  16.32  
6               8.27                   9.22  
10              7.58                   8.45  
2               4.45                   4.96  
0               4.02                   4.48  
5               3.06                   3.41  
1               2.10                   2.34  
4               1.17                   1.30  
3               0.06                   0.07  
8               0.01                   0.01

### Insights 

####  Less affluence
           
           unknown - 77% hex's contributing  26% gross 
           residential - 12% hex's contributing  25% gross
           transit_station - 1% hex's contributing  25% gross
            
####    High affluence
            
            residential - 40.77% hex's contributing  50% gross 
            office - 6% hex's contributing  16% gross
            leisure - 6% hex's contributing  9% gross  

In [27]:
df_type2_analysis = df_aff_use_case_type2 \
                            .groupby(['affluence_tag', 'residential']) \
                            .agg(
                                no_hexes = pd.NamedAgg('pickup_location_hex_8', 'nunique'),
                                gross_orders = pd.NamedAgg('total_gross_orders', 'sum'),
                                net_orders = pd.NamedAgg('total_net_orders', 'sum')
                                ).reset_index()

df_type2_analysis['city_hex_distr'] = (df_type2_analysis['no_hexes']*100.0/df_type2_analysis.no_hexes.sum()).round(2)
df_type2_analysis['city_gross_distr'] = (df_type2_analysis['gross_orders']*100.0/df_type2_analysis.gross_orders.sum()).round(2)
df_type2_analysis['agg_hexes'] = np.where(df_type2_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_hex_count.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_hex_count.sum()
                                         )
df_type2_analysis['agg_gross'] = np.where(df_type2_analysis['affluence_tag'] == 'High',
                                         df_aff_agg[(df_aff_agg['affluence_tag'] == 'High') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_gross_orders.sum(),
                                          df_aff_agg[(df_aff_agg['affluence_tag'] == 'Less') & (df_aff_agg['flag'] == 'gross_level_hex_tag')].aff_gross_orders.sum()
                                         )
df_type2_analysis['affluence_hex_distr'] = (df_type2_analysis['no_hexes']*100.0/df_type2_analysis['agg_hexes']).round(2)
df_type2_analysis['affluence_gross_distr'] = (df_type2_analysis['gross_orders']*100.0/df_type2_analysis['agg_gross']).round(2)


df_type2_analysis[['affluence_tag', 'residential', #'no_hexes',
                   'city_hex_distr', 'affluence_hex_distr',
                   'city_gross_distr', 'affluence_gross_distr']] \
.sort_values(['affluence_tag', 'residential'], ascending = False)

affluence_tag      residential  city_hex_distr  affluence_hex_distr  \
5          Less      residential            6.58                12.28   
4          Less  non-residential            5.33                 9.96   
3          Less          Unknown           41.64                77.76   
2          High      residential           18.93                40.77   
1          High  non-residential           12.71                27.37   
0          High          Unknown           14.80                31.87   

   city_gross_distr  affluence_gross_distr  
5              2.55                  24.78  
4              5.02                  48.83  
3              2.71                  26.39  
2             44.37                  49.45  
1             41.33                  46.07  
0              4.02                   4.48